File: tests/test_dispatch_engine.py
"""Tests for dispatch_engine.py — end-to-end, safety invariants, and bundle verification."""

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

import unittest
from harness_core.dispatch.dispatch_engine import DispatchEngine
from harness_core.dispatch.dispatch_ledger import DispatchBundle
from harness_core.dispatch.executor_adapter import ManualExecutor, MockExecutor, NoopExecutor

FIXTURES_DIR = Path(__file__).resolve().parent / "fixtures" / "dispatch"


class DispatchEngineBasicTests(unittest.TestCase):
    def setUp(self):
        self.engine = DispatchEngine()

    def test_dispatch_returns_bundle(self):
        bundle = self.engine.dispatch("Summarize the README")
        self.assertIsInstance(bundle, DispatchBundle)
        self.assertTrue(bundle.record.dispatch_id)
        self.assertEqual(bundle.record.final_status, "not_executed")

    def test_dispatch_with_mock_executor(self):
        engine = DispatchEngine(executor=MockExecutor())
        bundle = engine.dispatch("Summarize the README")
        self.assertEqual(bundle.record.final_status, "completed")
        self.assertEqual(bundle.execution_result.executor_type, "mock")
        self.assertIsNotNone(bundle.execution_result.output)

    def test_dispatch_with_manual_executor(self):
        engine = DispatchEngine(executor=ManualExecutor())
        bundle = engine.dispatch("Summarize the README")
        self.assertEqual(bundle.record.final_status, "manual_pending")
        self.assertIsNotNone(bundle.execution_result.prompt_pack)

    def test_dispatch_populates_record_fields(self):
        bundle = self.engine.dispatch("Test request")
        self.assertTrue(bundle.record.task_analysis_id)
        self.assertTrue(bundle.record.decision_id)
        self.assertTrue(bundle.record.budget_reservation_id)
        self.assertTrue(bundle.record.created_at)
        self.assertTrue(bundle.record.updated_at)

    def test_dispatch_stores_in_ledger(self):
        bundle = self.engine.dispatch("Test request")
        stored = self.engine._ledger.get_record(bundle.record.dispatch_id)
        self.assertIsNotNone(stored)
        self.assertEqual(stored.dispatch_id, bundle.record.dispatch_id)

    def test_dispatch_bundle_contains_full_chain(self):
        engine = DispatchEngine(executor=MockExecutor())
        bundle = engine.dispatch("Summarize the README")
        self.assertIsNotNone(bundle.analysis)
        self.assertIsNotNone(bundle.decision)
        self.assertIsNotNone(bundle.execution_result)
        self.assertIsNotNone(bundle.evaluation_result)
        self.assertEqual(bundle.analysis.analysis_id, bundle.decision.analysis_id)
        self.assertEqual(bundle.execution_result.dispatch_id, bundle.record.dispatch_id)
        self.assertEqual(bundle.evaluation_result.dispatch_id, bundle.record.dispatch_id)

    def test_dispatch_bundle_stored_in_ledger(self):
        engine = DispatchEngine(executor=MockExecutor())
        bundle = engine.dispatch("Test request")
        stored_bundle = engine._ledger.get_bundle(bundle.record.dispatch_id)
        self.assertIsNotNone(stored_bundle)
        self.assertEqual(stored_bundle.record.dispatch_id, bundle.record.dispatch_id)
        self.assertEqual(stored_bundle.analysis.analysis_id, bundle.analysis.analysis_id)

    def test_bundle_to_dict_roundtrip(self):
        engine = DispatchEngine(executor=MockExecutor())
        bundle = engine.dispatch("Test request")
        d = bundle.to_dict()
        self.assertIn("record", d)
        self.assertIn("analysis", d)
        self.assertIn("decision", d)
        self.assertIn("execution_result", d)
        self.assertIn("evaluation_result", d)


class SafetyInvariantTests(unittest.TestCase):
    def setUp(self):
        self.engine = DispatchEngine()

    def test_no_provider_executor_type(self):
        """Phase 1: provider executor must never be used."""
        bundle = self.engine.dispatch("Test request")
        self.assertNotEqual(bundle.execution_result.executor_type, "provider")

    def test_decision_has_shadow_routes_or_reason(self):
        bundle = self.engine.dispatch("Summarize the README")
        d = bundle.decision
        self.assertTrue(d.shadow_routes or d.no_shadow_route_reason)

    def test_budget_exists_before_execution(self):
        bundle = self.engine.dispatch("Test request")
        self.assertIsNotNone(bundle.record.budget_reservation_id)
        self.assertIsNotNone(bundle.decision.budget_reservation)

    def test_execution_result_links_to_record(self):
        engine = DispatchEngine(executor=MockExecutor())
        bundle = engine.dispatch("Test request")
        self.assertEqual(bundle.execution_result.dispatch_id, bundle.record.dispatch_id)
        self.assertEqual(bundle.execution_result.decision_id, bundle.decision.decision_id)

    def test_evaluation_result_links_to_record(self):
        engine = DispatchEngine(executor=MockExecutor())
        bundle = engine.dispatch("Test request")
        self.assertEqual(bundle.evaluation_result.dispatch_id, bundle.record.dispatch_id)
        self.assertEqual(bundle.evaluation_result.execution_result_id,
                         bundle.execution_result.result_id)

    def test_provider_disabled_gate_always_present(self):
        bundle = self.engine.dispatch("Test request")
        gate_types = [g.gate_type for g in bundle.decision.execution_gates]
        self.assertIn("provider_disabled", gate_types)
        self.assertIn("sandbox_disabled", gate_types)

    def test_low_risk_noop_is_decided(self):
        """Low-risk noop dispatch should have decision_status=decided."""
        bundle = self.engine.dispatch("Summarize the README")
        self.assertEqual(bundle.decision.decision_status, "decided")
        self.assertEqual(bundle.record.final_status, "not_executed")

    def test_low_risk_noop_gates_are_info(self):
        """Provider/sandbox gates for low-risk noop should be info, not block."""
        bundle = self.engine.dispatch("Summarize the README")
        for g in bundle.decision.execution_gates:
            if g.gate_type in ("provider_disabled", "sandbox_disabled"):
                self.assertEqual(g.severity, "info")


class GateTests(unittest.TestCase):
    def setUp(self):
        self.engine = DispatchEngine()

    def test_high_risk_generates_risk_gate(self):
        bundle = self.engine.dispatch("Fix the bug and commit changes to main")
        gate_types = [g.gate_type for g in bundle.decision.execution_gates]
        self.assertIn("target_write", gate_types)
        self.assertEqual(bundle.decision.decision_status, "needs_approval")

    def test_target_write_generates_gate(self):
        bundle = self.engine.dispatch("Fix the bug and commit the changes to main")
        gate_types = [g.gate_type for g in bundle.decision.execution_gates]
        self.assertIn("target_write", gate_types)

    def test_low_confidence_generates_gate(self):
        bundle = self.engine.dispatch("Make it better")
        gate_types = [g.gate_type for g in bundle.decision.execution_gates]
        self.assertIn("confidence", gate_types)


class GoldenFixtureE2ETests(unittest.TestCase):
    """End-to-end dispatch of all 20 golden fixtures."""

    def setUp(self):
        self.engine = DispatchEngine()

    def _run_fixture(self, filename):
        path = FIXTURES_DIR / filename
        with open(path) as f:
            fixture = json.load(f)

        bundle = self.engine.dispatch(
            fixture["raw_request"],
            request_source=fixture.get("request_source", "test_fixture"),
        )

        self.assertTrue(bundle.record.dispatch_id, f"dispatch_id missing for {filename}")
        self.assertTrue(bundle.record.task_analysis_id, f"task_analysis_id missing for {filename}")
        self.assertTrue(bundle.record.decision_id, f"decision_id missing for {filename}")
        self.assertIn(bundle.record.final_status,
                      ("not_executed", "completed", "failed", "escalated", "manual_pending"))
        return bundle

    def test_fixture_01(self):
        self._run_fixture("fixture_01_low_risk_summary.json")

    def test_fixture_02(self):
        self._run_fixture("fixture_02_doc_audit.json")

    def test_fixture_03(self):
        self._run_fixture("fixture_03_code_review.json")

    def test_fixture_04(self):
        self._run_fixture("fixture_04_code_gen.json")

    def test_fixture_05(self):
        self._run_fixture("fixture_05_debug.json")

    def test_fixture_06(self):
        self._run_fixture("fixture_06_architecture.json")

    def test_fixture_07(self):
        self._run_fixture("fixture_07_math.json")

    def test_fixture_08(self):
        self._run_fixture("fixture_08_config_review.json")

    def test_fixture_09(self):
        self._run_fixture("fixture_09_infra_deploy.json")

    def test_fixture_10(self):
        self._run_fixture("fixture_10_provider_boundary.json")

    def test_fixture_11(self):
        self._run_fixture("fixture_11_target_write.json")

    def test_fixture_12(self):
        self._run_fixture("fixture_12_secret_handling.json")

    def test_fixture_13(self):
        self._run_fixture("fixture_13_long_context.json")

    def test_fixture_14(self):
        self._run_fixture("fixture_14_ambiguous.json")

    def test_fixture_15(self):
        self._run_fixture("fixture_15_conflicting.json")

    def test_fixture_16(self):
        self._run_fixture("fixture_16_read_only_high_risk.json")

    def test_fixture_17(self):
        self._run_fixture("fixture_17_negated_no_write.json")

    def test_fixture_18(self):
        self._run_fixture("fixture_18_negated_no_execute.json")

    def test_fixture_19(self):
        self._run_fixture("fixture_19_budget_constrained.json")

    def test_fixture_20(self):
        self._run_fixture("fixture_20_high_quality_critical.json")


class DeterministicReplayTests(unittest.TestCase):
    def test_same_input_same_domain(self):
        engine = DispatchEngine()
        b1 = engine.dispatch("Summarize the README")
        b2 = engine.dispatch("Summarize the README")
        self.assertEqual(b1.analysis.task_domain, b2.analysis.task_domain)
        self.assertEqual(b1.analysis.task_intent, b2.analysis.task_intent)
        self.assertEqual(b1.record.final_status, b2.record.final_status)

    def test_bundle_analysis_matches_request(self):
        engine = DispatchEngine()
        bundle = engine.dispatch("Summarize the README")
        self.assertEqual(bundle.analysis.raw_request_snapshot, "Summarize the README")
        self.assertEqual(bundle.analysis.task_domain, "docs")
        self.assertEqual(bundle.analysis.task_intent, "summarize")


if __name__ == "__main__":
    unittest.main()

File: engine/tests/test_dispatch_engine.rs
use engine::dispatch_engine::DispatchEngine;

// Basic tests
#[test]
fn test_dispatch_returns_valid_json() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Summarize the README", "test_fixture");
    assert!(v.get("record").is_some());
    assert!(v.get("analysis").is_some());
    assert!(v.get("decision").is_some());
}

#[test]
fn test_dispatch_record_has_dispatch_id() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Summarize the README", "test_fixture");
    assert!(!v["record"]["dispatch_id"].as_str().unwrap().is_empty());
}

#[test]
fn test_dispatch_final_status_not_executed() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Summarize the README", "test_fixture");
    assert_eq!(v["record"]["final_status"], "not_executed");
}

#[test]
fn test_dispatch_populates_record_fields() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Test request", "test_fixture");
    assert!(!v["record"]["task_analysis_id"].as_str().unwrap().is_empty());
    assert!(!v["record"]["decision_id"].as_str().unwrap().is_empty());
    assert!(!v["record"]["budget_reservation_id"]
        .as_str()
        .unwrap()
        .is_empty());
}

#[test]
fn test_dispatch_bundle_contains_full_chain() {
    let engine = DispatchEngine::new();
    let bundle = engine.dispatch_bundle("Summarize the README", "test_fixture");
    let analysis_id = bundle.analysis["analysis_id"].as_str().unwrap();
    let decision_analysis_id = bundle.decision["analysis_id"].as_str().unwrap();
    assert_eq!(analysis_id, decision_analysis_id);
}

// Safety invariant tests
#[test]
fn test_no_provider_executor_type() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Test request", "test_fixture");
    assert_ne!(v["execution_result"]["executor_type"], "provider");
}

#[test]
fn test_multi_executor_without_cli_tier_reports_noop_policy() {
    let engine = DispatchEngine::with_multi_executor(engine::cli::MultiExecutor::new(
        std::collections::HashMap::new(),
    ));
    let v = engine.dispatch(
        "Generate code: function parseCsv(input: string): string[][].",
        "test_fixture",
    );

    assert_eq!(v["decision"]["selected_tier"], "codex_cli");
    assert_eq!(v["decision"]["execution_policy"]["executor_type"], "noop");
    assert_eq!(v["execution_result"]["executor_type"], "noop");
    let constraints = v["decision"]["hard_constraints"].as_array().unwrap();
    assert!(constraints
        .iter()
        .any(|value| value.as_str() == Some("no_provider_call")));
}

#[test]
fn test_decision_has_shadow_routes_or_reason() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Summarize the README", "test_fixture");
    let has_routes = v["decision"]["shadow_routes"]
        .as_array()
        .is_some_and(|a| !a.is_empty());
    let has_reason = v["decision"]["no_shadow_route_reason"].as_str().is_some();
    assert!(has_routes || has_reason);
}

#[test]
fn test_budget_exists_before_execution() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Test request", "test_fixture");
    assert!(v["record"]["budget_reservation_id"].as_str().is_some());
    assert!(v["decision"]["budget_reservation"].is_object());
}

#[test]
fn test_execution_result_links_to_record() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Test request", "test_fixture");
    assert_eq!(
        v["execution_result"]["dispatch_id"],
        v["record"]["dispatch_id"]
    );
    assert_eq!(
        v["execution_result"]["decision_id"],
        v["decision"]["decision_id"]
    );
}

#[test]
fn test_evaluation_result_links_to_record() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Test request", "test_fixture");
    assert_eq!(
        v["evaluation_result"]["dispatch_id"],
        v["record"]["dispatch_id"]
    );
    assert_eq!(
        v["evaluation_result"]["execution_result_id"],
        v["execution_result"]["result_id"]
    );
}

// Gate tests
#[test]
fn test_provider_disabled_gate_always_present() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Test request", "test_fixture");
    let gates = v["decision"]["execution_gates"].as_array().unwrap();
    let gate_types: Vec<&str> = gates
        .iter()
        .map(|g| g["gate_type"].as_str().unwrap())
        .collect();
    assert!(gate_types.contains(&"provider_disabled"));
    assert!(gate_types.contains(&"sandbox_disabled"));
}

#[test]
fn test_low_risk_noop_is_decided() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Summarize the README", "test_fixture");
    assert_eq!(v["decision"]["decision_status"], "decided");
    assert_eq!(v["record"]["final_status"], "not_executed");
}

#[test]
fn test_target_write_generates_gate() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Fix the bug and commit the changes to main", "test_fixture");
    let gates = v["decision"]["execution_gates"].as_array().unwrap();
    let gate_types: Vec<&str> = gates
        .iter()
        .map(|g| g["gate_type"].as_str().unwrap())
        .collect();
    assert!(gate_types.contains(&"target_write"));
}

#[test]
fn test_low_confidence_generates_gate() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Make it better", "test_fixture");
    let gates = v["decision"]["execution_gates"].as_array().unwrap();
    let gate_types: Vec<&str> = gates
        .iter()
        .map(|g| g["gate_type"].as_str().unwrap())
        .collect();
    assert!(gate_types.contains(&"confidence"));
}

#[test]
fn test_high_risk_needs_approval() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Fix the bug and commit changes to main", "test_fixture");
    assert_eq!(v["decision"]["decision_status"], "needs_approval");
}

// Deterministic replay tests
#[test]
fn test_same_input_same_domain() {
    let engine = DispatchEngine::new();
    let v1 = engine.dispatch("Summarize the README", "test_fixture");
    let v2 = engine.dispatch("Summarize the README", "test_fixture");
    assert_eq!(v1["analysis"]["task_domain"], v2["analysis"]["task_domain"]);
    assert_eq!(v1["analysis"]["task_intent"], v2["analysis"]["task_intent"]);
    assert_eq!(v1["record"]["final_status"], v2["record"]["final_status"]);
}

#[test]
fn test_bundle_analysis_matches_request() {
    let engine = DispatchEngine::new();
    let v = engine.dispatch("Summarize the README", "test_fixture");
    assert_eq!(
        v["analysis"]["raw_request_snapshot"],
        "Summarize the README"
    );
    assert_eq!(v["analysis"]["task_domain"], "docs");
    assert_eq!(v["analysis"]["task_intent"], "summarize");
}

// Golden fixture e2e tests (20 tests)
macro_rules! golden_e2e {
    ($name:ident, $file:expr) => {
        #[test]
        fn $name() {
            let engine = DispatchEngine::new();
            let path = Path::new(env!("CARGO_MANIFEST_DIR"))
                .parent()
                .expect("engine crate has workspace parent")
                .join("tests/fixtures/dispatch_wire/v1")
                .join($file);
            let raw = std::fs::read_to_string(&path).expect("fixture readable");
            let fixture: serde_json::Value = serde_json::from_str(&raw).expect("fixture json");
            let request = &fixture["request"];
            let raw_request = request["raw_request"].as_str().unwrap();
            let request_source = request["request_source"].as_str().unwrap_or("test_fixture");
            let v = engine.dispatch(raw_request, request_source);
            assert!(!v["record"]["dispatch_id"].as_str().unwrap().is_empty());
            assert!(!v["record"]["task_analysis_id"].as_str().unwrap().is_empty());
            let status = v["record"]["final_status"].as_str().unwrap();
            assert!([
                "not_executed",
                "completed",
                "failed",
                "escalated",
                "manual_pending"
            ]
            .contains(&status));
        }
    };
}

use std::path::Path;

golden_e2e!(e2e_fixture_01, "fixture_01_low_risk_summary.json");
golden_e2e!(e2e_fixture_02, "fixture_02_doc_audit.json");
golden_e2e!(e2e_fixture_03, "fixture_03_code_review.json");
golden_e2e!(e2e_fixture_04, "fixture_04_code_gen.json");
golden_e2e!(e2e_fixture_05, "fixture_05_debug.json");
golden_e2e!(e2e_fixture_06, "fixture_06_architecture.json");
golden_e2e!(e2e_fixture_07, "fixture_07_math.json");
golden_e2e!(e2e_fixture_08, "fixture_08_config_review.json");
golden_e2e!(e2e_fixture_09, "fixture_09_infra_deploy.json");
golden_e2e!(e2e_fixture_10, "fixture_10_provider_boundary.json");
golden_e2e!(e2e_fixture_11, "fixture_11_target_write.json");
golden_e2e!(e2e_fixture_12, "fixture_12_secret_handling.json");
golden_e2e!(e2e_fixture_13, "fixture_13_long_context.json");
golden_e2e!(e2e_fixture_14, "fixture_14_ambiguous.json");
golden_e2e!(e2e_fixture_15, "fixture_15_conflicting.json");
golden_e2e!(e2e_fixture_16, "fixture_16_read_only_high_risk.json");
golden_e2e!(e2e_fixture_17, "fixture_17_negated_no_write.json");
golden_e2e!(e2e_fixture_18, "fixture_18_negated_no_execute.json");
golden_e2e!(e2e_fixture_19, "fixture_19_budget_constrained.json");
golden_e2e!(e2e_fixture_20, "fixture_20_high_quality_critical.json");
